### DISCUSSION TITLE
Typescript needs to export all icon names in one type

### DISCUSSION THREAD

**Discussion by corysimmons** [tailwindlabs/heroicons#813] [Help]
- https://github.com/tailwindlabs/heroicons/issues/64
- similar issue: https://github.com/callstack/react-native-paper/issues/1475

I want to create a component called `<Icon />` that accepts a `name` prop that maps to a HeroIcon name.

**Comment by RobinMalfait** [comment]
This is typically not recommended because that means that all icons will exist in your bundle if you are purely relying on tree-shaking.

You can get the component names yourself by using something like this:
```ts
import * as Heroicons from '@heroicons/react/20/solid'
type Icons = keyof typeof Heroicons
```

**Reply by corysimmons** [reply]
>This is typically not recommended because that means that all icons will exist in your bundle if you are purely relying on tree-shaking.

All icons get added to your bundle if you are simply importing the icon name dictionary type like so?

```ts
export type HeroIconName = 'AcademicCap' | 'AdjustmentsHorizontal' // | ...
```

and then importing like:

```tsx
import { HeroIconName } from '@heroicons/react'

type Props = {
  name: HeroIconName
}

export default function Icon({name}: Props) {
  return <span>{name}</span>
}
```

**Reply by corysimmons** [reply]
If this works without adding every icon to your bundle:

```ts
import * as Heroicons from '@heroicons/react/20/solid'
type Icons = keyof typeof Heroicons
```

...can we just export that type as part of the HeroIcons npm package for dev convenience?

---

Actually, _this_ ☝️ implementation seems like it would add every icon to the bundle. The string implementation (i.e. `export type HeroIconName = 'AcademicCap' | 'AdjustmentsHorizontal' // | ...`) doesn't seem like it would bundle every icon.
